apps/src/lib/node/ledger/shell/vote_extensions.rs

```diff
@@ -0,0 +1,460 @@
+#[cfg(not(feature = "ABCI"))]
+mod extend_votes {
+    use anoma::types::ethereum_events::vote_extensions::{
+        EpochPower, SignedEthEvent, SignedEvent,
+    };
+
+    use super::super::*;
+
+    /// The data we include in a vote extension
+    #[derive(Clone, Debug, BorshSerialize, BorshDeserialize)]
+    pub struct VoteExtension {
+        /// Ethereum events seen since last round
+        ethereum_events: Vec<SignedEthEvent>,
+    }
+
+    impl<D, H> Shell<D, H>
+    where
+        D: DB + for<'iter> DBIter<'iter> + Sync + 'static,
+        H: StorageHasher + Sync + 'static,
+    {
+        /// INVARIANT: This method must be stateless.
+        pub fn extend_vote(
+            &mut self,
+            _req: request::ExtendVote,
+        ) -> response::ExtendVote {
+            response::ExtendVote {
+                vote_extension: VoteExtension {
+                    ethereum_events: self.new_ethereum_events(),
+                }
+                .try_to_vec()
+                .unwrap(),
+            }
+        }
+
+        /// At present this checks the signature on all Ethereum headers
+        ///
+        /// INVARIANT: This method must be stateless.
+        pub fn verify_vote_extension(
+            &self,
+            req: request::VerifyVoteExtension,
+        ) -> response::VerifyVoteExtension {
+            if let Ok(VoteExtension { ethereum_events }) =
+                VoteExtension::try_from_slice(&req.vote_extension[..])
+            {
+                return if ethereum_events.iter().all(|event| {
+                    self.validate_ethereum_event(
+                        self.storage.last_height + 1,
+                        event,
+                    )
+                }) {
+                    response::VerifyVoteExtension {
+                        status: VerifyStatus::Accept.into(),
+                    }
+                } else {
+                    response::VerifyVoteExtension {
+                        status: VerifyStatus::Reject.into(),
+                    }
+                };
+            }
+            Default::default()
```

Comment:
Been looking more at how vote extensions work - the `Default::default()` here will return [`VerifyStatus::Unknown`](https://docs.tendermint.com/master/spec/abci++/abci++_methods.html#verifystatus) which will panic Tendermint (we can hit this code path if a validator's vote extension doesn't serialize to a `VoteExtension { ethereum_events }`). We should explicitly accept or reject in the case where that happens.

I think we should reject the validator's bad vote now although I'm not 100% on that.

Although Tendermint docs says we SHOULD accept invalid vote extensions at this stage and deal with them in our logic later, we could do this change later on (accept syntactically invalid vote extensions and the block proposer can flag it as invalid in `PrepareProposal`).